File: C++/initial_hamiltonian.cpp
#include "initial_hamiltonian.h"

void add(Matrix* H, int x, int y) {
  (*H).append(x , y, -1);

  return;
}

void recurse(Matrix* H, int global_row, int row, int col_offset, int n) {
    bool up = row % 2 == 0;
    int dim = 1 << n;
    
    if (n == 1) {
        if (up) add(H, global_row, col_offset + 1);
        else add(H, global_row, col_offset);
    }
    else {
        int half = dim / 2;
        int ident_col = row % half;
        // Top half
        if (row < half) {
            recurse(H, global_row, row, col_offset, n-1);
            add(H, global_row, col_offset + ident_col + half);
        }
        // Bottom half
        else {
            add(H, global_row, col_offset + ident_col);
            recurse(H, global_row, row % half, col_offset + half, n-1);
        }
    }
    
    return;
}

Matrix initial_hamiltonian(int n) {
  int dim = pow(2, n);
  int n_non_zero = n * dim;

  Matrix H(dim, dim);

  H.reserve(n_non_zero);

  for (int i = 0; i < dim; i++) {
    //H.reserve(n, i);
    recurse(&H, i, i, 0, n);
    H.finalize(i);
  }

  return H;
}

File: matlab/bad/InitialHamiltonian.m
% Optimized (Recursive Version)

% Returns a sparse matrix
function H = InitialHamiltonian(N)
    dim = 2^N;
    H = sparse([], [], [], dim, dim);

    for k = 1:dim
        H = recurse(H3, k, k-1, 1, N);
    end
end

function H = recurse(H, global_row, row, col_offset, n)
    dim = 2 ^ n;

    if n == 1
        if mod(row, 2) == 0
            H(global_row, col_offset + 1) = -1;
        else
            H(global_row, col_offset) = -1;
        end
    else
        half = dim / 2;
        ident_col = mod(row, half);

        if row < half
            H = recurse(H, global_row, row, col_offset, n-1);
            H(global_row, col_offset + ident_col + half) = -1;
        else
            H(global_row, col_offset + ident_col) = -1;
            H = recurse(H, global_row, ident_col, col_offset + half, n-1);
        end
    end
end